src/vmime/net/transport.hpp

```diff
@@ -69,10 +69,16 @@ class VMIME_EXPORT transport : public service {
 	  *
 	  * @param msg message to send
 	  * @param progress progress listener, or NULL if not used
+	  * @param dsnNotify comma separated list of notification conditions as specified in RFC 1891
+	  * @param dsnRet content of DSN - full message or headers only ("FULL" or "HDRS")
+	  * @param dsnEnvelopId envelop ID to be able to pair the DSN with original message
 	  */
 	virtual void send(
 		const shared_ptr <vmime::message>& msg,
-		utility::progressListener* progress = NULL
+		utility::progressListener* progress = NULL,
+		const std::string& dsnNotify = std::string(),
+		const std::string& dsnRet = std::string(),
+		const std::string& dsnEnvelopId = std::string()
```

Comment:
Better use an object to encapsulate DSN settings: declare a class in a separate file, and use a const ref here (and other send() prototypes). Also, better use a std::vector<> for notification events.

Reply:
OK, I have created a simple "dsnAttributes" class. I have left all three attributes as text for now as they are used everywhere as-is.